Mathlib/Geometry/Manifold/ClosedNManifold.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2024 Michael Rothgang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Rothgang
+-/
+import Mathlib.Geometry.Manifold.Instances.Real
+import Mathlib.Geometry.Manifold.Instances.Sphere
+import Mathlib.Geometry.Manifold.InteriorBoundary
+
+/-!
+# Closed `n`-dimensional manifolds
+
+We define closed `n`-dimensional manifolds and show a few basic properties.
+
+## Main definitions
+* `ClosedManifold M I`: a topological manifold `M` is closed if it is compact and boundaryless
+* `NManifold n M I`: an n-manifold is a smooth `n`-dimensional manifold `M`.
+* `ClosedNManifold n M I`: a closed n-manifold `M` is both closed and an `n`-manifold
+
+## Main results
+* `ClosedManifold.prod`: the product of closed manifolds is closed
+* `NManifold.prod`: the product of an `n`-manifold and an `m`-manifold is an `n+m`-manifold
+* `ClosedNManifold.prod`: the product of a closed `n`-manifold and a closed `m`-manifold is a closed
+`n+m`-manifold
+* We prove a few basic examples
+  - the empty set is closed and an `n`-manifold, for every `n` (TODO insert!)
+  - Euclidean n-space is an n-manifold
+  - the standard n-sphere is a closed n-manifold
+  - the standard two-torus `S¹ × S¹` is a closed 2-manifold
+
+TODO: investigate why the spheres fail!
+-/
+
+open scoped Manifold
+open Metric (sphere)
+open FiniteDimensional Set
+
+noncomputable section
+
+variable (n : ℕ) {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  -- declare a smooth manifold `M` over the pair `(E, H)`.
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E] {H : Type*} [TopologicalSpace H]
+  (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+  (I : ModelWithCorners 𝕜 E H) [SmoothManifoldWithCorners I M]
+
+/-- A topological manifold is called **closed** iff it is compact without boundary. -/
+structure ClosedManifold [CompactSpace M] [BoundarylessManifold I M]
+
+variable {E' : Type*} [NormedAddCommGroup E'] [NormedSpace 𝕜 E']
+  {H' : Type*} [TopologicalSpace H'] (N : Type*) [TopologicalSpace N] [ChartedSpace H' N]
+  (J : ModelWithCorners 𝕜 E' H') [SmoothManifoldWithCorners J N]
+
+instance ClosedManifold.prod [CompactSpace M] [BoundarylessManifold I M]
+    [CompactSpace N] [BoundarylessManifold J N] :
+  ClosedManifold (M × N) (I.prod J) where
+
+/-- An **n-manifold** is a smooth `n`-dimensional manifold. -/
+structure NManifold [NormedAddCommGroup E]  [NormedSpace 𝕜 E] [FiniteDimensional 𝕜 E]
```

Comment:
Again, I don't really see this as a good design. Instead, you should probably use an assumption `[Fact (finrank 𝕜 E = n)]` to say that the manifold is `n`-dimensional.